state/machine.go

```diff
@@ -2076,10 +2079,28 @@ func (m *Machine) CreateUpgradeSeriesLock() error {
 			return nil, errors.AlreadyExistsf("upgrade series lock for machine %q", m)
 		}
 		if err = m.isStillAlive(); err != nil {
-			return nil, errors.Wrap(jujutxn.ErrNoOperations, err)
+			return nil, errors.Trace(err)
+		}
+		// Exit early if the Machine series doesn't need to change.
+		fromSeries := m.Series()
+		if fromSeries == toSeries {
+			return nil, errors.Trace(errors.Errorf("machine %s already at series %s", m.Id(), toSeries))
```

Comment:
I was wondering if there was a way to wrap these kinds of errors so as to have both the effect of the no transaction errors while still propagating up the relevant error message.

Reply:
I was wondering the same thing, but haven't found one, there's a comment in txn.Run()
if len(ops) == 0 {
			// Treat this the same as ErrNoOperations but don't suppress other errors.
			return nil
}
However err == ErrNoOperations, returns nil also, not sure how it doesn't suppress other errors.

I think we'll have to some error handling on the call. Adding a todo in machinemanager.

Reply:
Do you need to trace an error at its source?